### PR TITLE
feat: loading documents using `from_files`

### LINKED ISSUE #1165: support loading documents with `from_files` signature
**Is your feature request related to a problem? Please describe.**

In docarray v1, we support:

```python
da = DocumentArray.from_files('imgs/*.jpg')
```

While it is missing in the current v2. Could be very useful for us to loading the documents.

### PR DESCRIPTION
**Pull request by agaraman0**
Goals:

- issue fix #1165 
- ...
```python
import docarray
da = docarray.DocumentArray.from_files("path_to_dir/*.jpg")
```
Todo

- [x] function implementation
- [ ] function tests
- [ ] documentation changes
- [ ] function args relevance

- [ ] check and update documentation, if required. See [guide](https://github.com/jina-ai/jina/blob/master/CONTRIBUTING.md#documentation-guidelines)

### CODE REVIEW

**Comment by samsja** [comment]
Hey @agaraman0 can u update the issue description with more detail on the PR ? Maybe add a code snippet of the new API

File: docarray/array/array/io.py [hunk 1]
```diff
@@ -302,6 +306,75 @@ def to_json(self) -> str:
...
+        num_docs = 0
+        if isinstance(patterns, str):
+            patterns = [patterns]
+
+        _r = None
+        if exclude_regex:
+            try:
+                _r = re.compile(exclude_regex)
+            except re.error:
+                raise ValueError(f'`{exclude_regex}` is not a valid regex.')
```
Comment (samsja) [inline docarray/array/array/io.py:358]: lets show the original message from regex as well. Can be useful for debugging

File: docarray/array/array/io.py [hunk 2]
```diff
@@ -302,6 +306,75 @@ def to_json(self) -> str:
...
+                    d = BaseDocument(uri=g)
+                    document_array.append(d)
+                elif read_mode in {'r', 'rb'}:
+                    with open(g, read_mode) as fp:
+                        d = BaseDocument(content=fp.read(), uri=g)
+                        document_array.append(d)
+                num_docs += 1
+            if size is not None and num_docs >= size:
+                break
+        return DocumentArray[BaseDocument](document_array)
```
Comment (samsja) [inline docarray/array/array/io.py:376]:
```suggestion
        return cls(document_array)
```

Here it should return a documentarray of the same cls

File: docarray/array/array/io.py [hunk 3]
```diff
@@ -302,6 +306,75 @@ def to_json(self) -> str:
...
+        cls,
+        patterns: Union[str, List[str]],
+        recursive: bool = True,
+        size: Optional[int] = None,
+        sampling_rate: Optional[float] = None,
+        read_mode: Optional[str] = None,
+        exclude_regex: Optional[str] = None,
+        *args,
+        **kwargs,
+    ) -> 'DocumentArray':
```
Comment (samsja) [inline docarray/array/array/io.py:320]: can u use TypeVar to higlight than the return document will have the same type as `cls` ?

Usually it is does by doing:

```
@classmethod
def my_method(cls: Type[T]) -> T:
   ...
```

File: docarray/array/array/io.py [hunk 4]
```diff
@@ -302,6 +306,75 @@ def to_json(self) -> str:
...
+        :param exclude_regex: if set, then filenames that match to this pattern are not included.
+        :yield: file paths or binary content
+        .. note::
+            This function should not be directly used, use :meth:`Flow.index_files`, :meth:`Flow.search_files` instead
+        """
+        from docarray import DocumentArray
+
+        document_array = []
+        if read_mode not in {'r', 'rb', None}:
+            raise RuntimeError(
```
Comment (samsja) [inline docarray/array/array/io.py:340]:
```suggestion
            raise ValueError(
```

File: docarray/array/array/io.py [hunk 5]
```diff
@@ -302,6 +306,75 @@ def to_json(self) -> str:
...
+                _r = re.compile(exclude_regex)
+            except re.error:
+                raise ValueError(f'`{exclude_regex}` is not a valid regex.')
+
+        for g in _iter_file_exts(patterns):
+            if os.path.isdir(g):
+                continue
+            if _r and _r.match(g):
+                continue
+            if sampling_rate is None or random.random() < sampling_rate:
```
Comment (samsja) [inline docarray/array/array/io.py:365]: what is this sampling rate ?

File: docarray/array/array/io.py [hunk 6]
```diff
@@ -302,6 +306,75 @@ def to_json(self) -> str:
...
+                if read_mode is None:
+                    d = BaseDocument(uri=g)
+                    document_array.append(d)
+                elif read_mode in {'r', 'rb'}:
+                    with open(g, read_mode) as fp:
+                        d = BaseDocument(content=fp.read(), uri=g)
```
Comment (samsja) [inline docarray/array/array/io.py:371]: we don't have this `content` anymore in BaseDocument. Please refer to the `from_csv` classmethod to see how it should be handled now

**Review by samsja — CHANGES_REQUESTED** [review]
Thanks for the work :rocket: I added a couple of comments

File: docarray/array/array/io.py (hunk 5 above)
Comment (agaraman0) [reply docarray/array/array/io.py:364]: not sure about the relevance of `sampling_rate`, I have added a refactored version of `from_files` from the main branch, and tested over common functionality behavior that worked out fine, still have some doubts will fix them.

**Comment by samsja** [comment]
closed this PR as the issue was solve in another PR

**Pull request closed without merging**